go/libkb/log_send.go

```diff
@@ -239,6 +243,81 @@ func findFirstNewline(b []byte) []byte {
 	return b[(index + 1):]
 }
 
+func appendError(collected []byte, err error) []byte {
+	return append(collected, []byte(fmt.Sprintf("\nERROR READING LOGS:\n%#v\n", err))...)
+}
+
+func readBytesFromPipe(log logger.Logger, pipe io.ReadCloser, numBytes int) ([]byte, error) {
+	buf := [4096]byte{}
+	collected := []byte{}
+	for {
+		n, err := pipe.Read(buf[:])
+		collected = append(collected, buf[:n]...)
+		if err == io.EOF {
+			_ = pipe.Close()
+			return collected, nil
+		}
+		if err != nil {
+			// For non-EOF errors, append them to the logs and return what we have.
+			return appendError(collected, err), err
+		}
+		// If the output we've collected grows bigger than it's supposed to be,
+		// just drop everything we've seen so far, add a big warning at the
+		// top, and keep going. This avoids consuming all the memory on the
+		// system if the logs are pathological, but still lets us see the tail
+		// end of the logs, which is probably what we care about.
+		if len(collected) > numBytes {
+			msg := "WARNING: Lines in the journal are longer than expected. Truncating log."
+			log.Warning(msg)
+			collected = []byte("#####\n" + msg + "\n#####\n")
+		}
+	}
+}
+
+// If systemd is running, we look for logs in the systemd journal for our 3
+// main services.
+func tailSystemdJournal(log logger.Logger, which string, numBytes int) (ret string) {
+	log.Debug("+ tailing journalctl for %s (%d bytes)", which, numBytes)
+	defer func() {
+		log.Debug("- scanned %d bytes", len(ret))
+	}()
+
+	// Journalctl doesn't provide a "last N bytes" flag directly, so instead we
+	// use "last N lines". My own logs for all three services average out to
+	// between 120 and 220 bytes per line. We'll use a slightly conservative
+	// assumption of 300 bytes per line.
+	numLines := numBytes / 300
```

Comment:
I think we should change 300 to 150 so that 16MB / 150 gives about the number of lines that currently show up in service logs.

Reply:
Will do, like we talked about in chat. I'll add a rule like "if we go over 200% of what we expected, then truncate".